### PR TITLE
BENCH: add benchmark for energy_distance and wasserstein_distance

### PR DESCRIPTION
**Pull request by charlotte12l**
#### What does this implement/fix?
Add benchmark for energy_distance and wasserstein_distance.

### CODE REVIEW

File: benchmarks/benchmarks/stats.py [hunk 1]
```diff
@@ -434,3 +434,16 @@ def time_fni(self, distcase):
...
+    def setup(self):
+        rng = np.random.default_rng(12345678)
+        self.u_values, self.u_weights = rng.random(4) * 10, rng.random(4) * 10
+        self.v_values, self.v_weights = rng.random(2) * 10, rng.random(2) * 10
+
+    def time_energy_distance(self):
+        distance = stats.energy_distance(self.u_values, self.v_values, self.u_weights, self.v_weights)
+
+    def time_wasserstein_distance(self):
+        distance = stats.wasserstein_distance(self.u_values, self.v_values, self.u_weights, self.v_weights)
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py:449]: Could you break this line to make it fit in 80 char line length? And same for the `energy_distance` call above?

File: benchmarks/benchmarks/stats.py [hunk 2]
```diff
@@ -434,3 +434,16 @@ def time_fni(self, distcase):
             sup.filter(RuntimeWarning, "divide by zero")
             sup.filter(RuntimeWarning, "invalid value encountered")
             stats.NumericalInverseHermite(dist)
+
+
+class DistanceFunctions(Benchmark):
+    def setup(self):
+        rng = np.random.default_rng(12345678)
+        self.u_values, self.u_weights = rng.random(4) * 10, rng.random(4) * 10
+        self.v_values, self.v_weights = rng.random(2) * 10, rng.random(2) * 10
```
Comment (rgommers) [inline benchmarks/benchmarks/stats.py:443]: It looks to me like the array sizes are so small here that you're only measuring function overhead. If I change it to:
```python
        size = 4000
        rng = np.random.default_rng(12345678)
        self.u_values = rng.random(size) * 10
        self.u_weights = rng.random(size) * 10
        self.v_values = rng.random(size // 2) * 10
        self.v_weights = rng.random(size // 2) * 10
```
then I see that for size `4` or `40` the benchmark takes about 40 us, for `400` it takes ~100 us, and for `4000` it takes ~800 us. So you may want to make this a lot larger (maybe ~10,000?). 

You'll also want to check that the size is such that the performance improvements with Pythran you're trying to show actually show up - in gh-14154 it looks like for very small arrays the gain is ~3x, but what about large arrays? Parametrizing the benchmark on size (e.g. one small size, one large one) may make sense.

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @charlotte12l! Sorry for the delay in reviewing. My main comment is about size of arrays to be benchmarked.

File: benchmarks/benchmarks/stats.py (hunk 2 above)
Comment (charlotte12l) [reply benchmarks/benchmarks/stats.py:443]: Thanks for reminding me about the size!  I just found the size is really a big problem... When I run the benchmark with `size = 4000`, the master branch is ~1.32ms while my branch is ~3.39ms. Seems in this case, a larger input size would make pythran slower than the original code... I just submitted an issue in Pythran https://github.com/serge-sans-paille/pythran/issues/1793

**Review by rgommers — CHANGES_REQUESTED** [review]
Using `10` and `4000` looks good to me. There's a better way to write this code though, using parametrization. See how `param_names` and `params` are used in https://github.com/scipy/scipy/blob/5b6816a9dacc64f82bb95051bcf9d2120030147e/benchmarks/benchmarks/stats.py#L235 for example. 

Doing that will also give you nicer benchmark output.

**Comment by charlotte12l** [comment]
Just changed to use parametrization.
```
[  0.00%] ·· Benchmarking existing-py_Users_charlotte_anaconda3_envs_scipydev_bin_python
[ 25.00%] ··· Running (stats.DistanceFunctions.time_energy_distance--)..
[ 75.00%] ··· stats.DistanceFunctions.time_energy_distance                                                                    ok
[ 75.00%] ··· ======== =============
               n_size               
              -------- -------------
                 10       54.4±6μs  
                4000    1.37±0.09ms 
              ======== =============

[100.00%] ··· stats.DistanceFunctions.time_wasserstein_distance                                                               ok
[100.00%] ··· ======== =============
               n_size               
              -------- -------------
                 10       57.0±6μs  
                4000    1.52±0.04ms 
              ======== =============
```

**Review by rgommers — APPROVED** [review]
LGTM now, thanks @charlotte12l

**Pull request merged**